paf-mvp-cmp/src/controller.ts

```diff
@@ -0,0 +1,565 @@
+    implementor will need to ensure they implement the interface. This approach is less flexible.
+    
+    The method then needs to return the values as they currently exist in the persistent storage. The caller is then
+    responsible for handling the result and the UI. We need to give more thought to the unhappy path here and how errors
+    will be handled and communicated to the user. The UI doesn't have placeholders to tell the user that something has
+    gone wrong. i.e. "Whoops. We're not able to store your preference at the moment. We'll store them just for this 
+    site, so that you can continue. Okay?".
+    
+    This call should become.
+    
+      updateIdsAndPreferences(
+        {
+          proxyHostName: this.config.proxyHostName
+        },
+        {
+          identifiers: [ array of identifiers ],
+          preferences: signedPreferences
+        }
+      ) : Promise<IdsAndOptionalPreferencesWithErrorCodes> 
+
+    For reference the SWAN API provided a single method for getting and updating the data. 
+    See https://github.com/SWAN-community/swan/blob/main/apis.md#update
+    SWID is similar to the paf_browser_id or Random ID (RID)
+    Pref is similar to the PreferencesData structure.
+    OWID is similar to Source.
+    The design approach there is to have a single method that will store what is provided (if anything) and return the
+    current data. The CMP would handle the decrypt of the results which is not relevant to PAF as the data is not
+    encrypted.
+    */
+
+    // Update the ids and preferences.
+    await updateIdsAndPreferences(
+      this.config.proxyHostName,
+      this.model.pref.persisted.data.use_browsing_for_personalization,
+      [this.model.rid.value]
+    );
+
+    // Refresh the ids and preferences.
+    const r = await refreshIdsAndPreferences({
+      proxyHostName: this.config.proxyHostName,
+      triggerRedirectIfNeeded: true,
+      showPrompt: ShowPromptOption.doNotPrompt,
+    });
+
+    return r.data;
+  }
+
+  /**
+   * If there are no identifiers then get a new one.
+   * @returns a new random identifier from the Operator
+   */
+  private getNewIdIfNeeded(): Promise<Identifier> {
+    if (
+      this.model.rid.value === null ||
+      this.model.rid.value.source === null ||
+      this.model.rid.value.source.signature === null
+    ) {
+      return this.resetId();
+    }
+    return Promise.resolve<Identifier>(this.model.rid.value);
+  }
+
+  /**
+   * Resets the random identifier by fetching a new one from the Operator.
+   * @returns
+   */
+  private resetId(): Promise<Identifier> {
+    return getNewId({
+      proxyHostName: this.config.proxyHostName,
+    });
+  }
+
+  /**
+   * Signs the preferences with the CMP provider if they have not been signed already.
+   * @returns signed preferences, which might be the same as the existing ones
+   */
+  private signIfNeeded(): Promise<Preferences> {
+    if (this.model.pref.hasChanged) {
```

Comment:
I believe there is a problem here:
- if at https://github.com/prebid/paf-mvp-implementation/pull/120/files?short_path=496599c#diff-21193811b758f6d40b39e174d4b5622f8b741a520255c5351f18056bf93fe4c8R416, the id is re-generated,  `this.model.pref.hasChanged` is not set to `true`, and preferences are not signed. This is an issue: preferences will be signed with a value of the id that not the current id